### ISSUE TITLE
The reason for NaN

### ISSUE THREAD

**Issue by KwangryeolPark** [ultralytics/yolov5#12591]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

Like other issues, I also see NaN during training yolov5m to coco dataset following the script in coco.yaml and README.md. 

I try to figure out the reason for NaN and I find a hint in a Issue which indirectly is about amp (Auto Mixed Precision).

It makes sense that low precission has a higher chance to occur NaN during casting because of Underflow.

Therefore, I think, lots of NaN problem come from amp so I looks better to use NVIDIA apex which uses distribution shift to prevent distribution miss match.

### Environment

YOLOv5m
torch:1.12.1+cu116
python: 3.8.12
dataset: coco
optimizer: CAME
epochs: 300
batch size: 40

### Minimal Reproducible Example

```
python train.py --data coco.yaml --epochs 300 --weights '' --cfg yolov5m.yaml --batch-size 40 --optimizer CAME --device 0
```

I use <a href="https://github.com/yangluo7/CAME">CAME</a> optimizer with betas=(momentum, 0.999, 0.999)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by KwangryeolPark** [comment]
I hope you fix the mixed precision problem.

**Comment by glenn-jocher** [comment]
@KwangryeolPark hello! Thanks for bringing this to our attention. NaNs during training can indeed sometimes be related to precision issues when using mixed precision training (AMP). However, there could be other factors at play, such as learning rate, weight initialization, or data preprocessing.

Regarding the use of NVIDIA apex, YOLOv5 uses PyTorch's native AMP implementation, which is generally recommended for its ease of use and integration. If you're experiencing NaNs with AMP, you might want to try the following:

1. Reduce the learning rate.
2. Increase the batch size if possible, as smaller batches can sometimes lead to instability with AMP.
3. Ensure that your data preprocessing is correct and that there are no anomalies in the dataset.
4. Experiment with different optimizers if the issue persists.

If you're willing to submit a PR, we'd be happy to review any improvements or fixes you propose. Just make sure to thoroughly test your changes to ensure they're beneficial across various scenarios.

Remember to check out our documentation for more details on troubleshooting and best practices: https://docs.ultralytics.com/yolov5/

Thanks for your contribution to the YOLOv5 community! 🚀

**Comment by KwangryeolPark** [comment]
@glenn-jocher Thank you for answer.

In order to set learning-rate, I see <a href="https://docs.ultralytics.com/modes/train/#arguments">Training Arguments </a> and find lr0 argument. However, when I add --lr0 0.001, the script shows `train.py: error: unrecognized arguments: --lr0 1e-3`.

**Comment by glenn-jocher** [comment]
Apologies for the confusion, @KwangryeolPark. The correct argument for setting the initial learning rate in the YOLOv5 training script is `--lr`. So, if you want to set the initial learning rate to 0.001, you should use the following command:

```bash
python train.py --data coco.yaml --epochs 300 --weights '' --cfg yolov5m.yaml --batch-size 40 --optimizer CAME --device 0 --lr 0.001
```

Make sure to adjust the learning rate according to your specific needs and keep an eye on the training process to ensure stability. If you have any further questions or issues, don't hesitate to reach out. Happy training! 🚀

**Comment by KwangryeolPark** [comment]
@glenn-jocher Thank you for guidance. However, `--lr 0.001` argument also occur: `train.py: error: unrecognized arguments: --lr 0.001`

**Comment by glenn-jocher** [comment]
I apologize for the oversight, @KwangryeolPark. In YOLOv5, the learning rate is set in the hyperparameter configuration file rather than as a command-line argument. You can adjust the learning rate by editing the `hyp.scratch.yaml` file or any other hyperparameter file you are using.

For example, to set the initial learning rate to 0.001, you would modify the `lr0` value in your hyperparameter file like so:

```yaml
lr0: 0.001  # initial learning rate
```

Then, you can reference this hyperparameter file during training using the `--hyp` argument:

```bash
python train.py --data coco.yaml --epochs 300 --weights '' --cfg yolov5m.yaml --batch-size 40 --optimizer CAME --device 0 --hyp your_hyperparameter_file.yaml
```

Replace `your_hyperparameter_file.yaml` with the path to your edited hyperparameter file. This should correctly set the initial learning rate for your training session. If you encounter any further issues, please let us know. Good luck with your training! 🌟

**Comment by KwangryeolPark** [comment]
Thank you

**Comment by glenn-jocher** [comment]
You're welcome, @KwangryeolPark! If you have any more questions or need further assistance in the future, feel free to reach out. Best of luck with your YOLOv5 training! Happy detecting! 🚀👀

**Comment by tobymuller233** [comment]
Hi! Thanks for your reply.
I originally trained my model normally using the framework, but everything changed after I attempted to fine-tune the pruned model. I previously pruned the model using the torch_pruning tool and then tried to fine-tune it through training in the framework, only to find that the loss would become NaN. After investigation, I identified that the issue might be with the BN layer, and modifying it temporarily solved my problem.

However, when I ran train.py again and trained the model that I had previously trained and tested successfully, the loss still resulted in an error. Using the same configuration and parameters with which I had previously trained the model successfully, it now always encounters NaN at the Upsample layer. I’m completely baffled by this situation.

**Comment by pderrenger** [comment]
Thank you for the detailed explanation of your issue. Since you're encountering NaN losses after model pruning and during retraining, this could be related to several factors:

1. Verify your learning rate isn't too high - try reducing it in your hyperparameter file (hyp.scratch.yaml) by setting a lower `lr0` value, such as 0.001 or 0.0001

2. Consider disabling AMP (mixed precision) training by adding `--no-amp` to your training command to see if precision issues are causing the NaNs

3. When fine-tuning pruned models, you may need to adjust the batch normalization layers' momentum and eps parameters, as pruning can affect their stability

Could you share your current training command and the specific version of YOLOv5 you're using? This would help us better diagnose the issue.

Also, consider checking our training guide for best practices: https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/

**Comment by tobymuller233** [comment]
Since in my situation, nan values appears after a nn.Upsample module, which means it's in the forwarding procedure, not backward. So I don't think this is related to lr, here is my model yaml:
```yaml
nc: 1 # number of classes
depth_multiple: 1.0 # model depth multiple
width_multiple: 1.0 # layer channel multiple
anchors:
  - [4, 6, 7, 10, 11, 15]
  - [16, 24, 33, 25, 26, 41]
  - [47, 60, 83, 97, 141, 149]

backbone:
  # [from, number, module, args]
  # args: out_channels, size, stride
  [
    [-1, 1, Conv, [1, 3, 2]],  # 0  [batch, 8, size/2, size/2]
    [-1, 1, DWConv, [1, 3, 1]], # 1 [320]
    [-1, 1, Conv, [6, 1, 1 ]], # 2  [320]
    [-1, 1, Conv, [14, 1, 1]], # 3 [-1, 1, DWConv, [24, 3, 2]] # 4
    [-1, 1, Conv, [4, 1, 1]], # 4
    [-1, 1, Bottleneck3, [4, 30]], # 5

    [-1, 1, Conv, [24, 1, 1]], # 6  
    [-1, 1, DWConv, [24, 3, 2]], # 7  [160]
    [-1, 1, Conv, [7, 1, 1]], # 8
    [-1, 1, Bottleneck3, [7, 26]], # 9
    [-1, 1, Bottleneck3, [7, 34]], # 10
    
    [-1, 1, Conv, [28, 1, 1]], # 11 
    [-1, 1, DWConv, [28, 3, 2]], # 12 [80] 
    [-1, 1, Conv, [9, 1, 1]], # 13
    [-1, 1, Bottleneck3, [9, 60]], # 14
    [-1, 1, Bottleneck3, [9, 71]], # 15
    [-1, 1, Bottleneck3, [9, 76]], # 16

    [-1, 1, Conv, [57, 1, 1]], # 17
    [-1, 1, DWConv, [57, 3, 1]], # 18
    [-1, 1, Conv, [20, 1, 1]], # 19
    [-1, 1, Bottleneck3, [20, 84]], # 20
    [-1, 1, Bottleneck3, [20, 111]], # 21

    [-1, 1, Conv, [86, 1, 1]], # 22    [80]
    [-1, 1, DWConv, [86, 3, 2]], # 23  [80] -> [40] 
    [-1, 1, Conv, [16, 1, 1]], # 24
    [-1, 1, Bottleneck3, [16, 92]], # 25 [batch, 40, size/16, size/16]
    [-1, 1, Bottleneck3, [16, 151]], # 26 [batch, 40, size/16, size/16]
  ]

head: [
    [-1, 1, Conv, [48, 1, 1]], # 27 [40]
    [[-1, -5], 1, Concat, [1]], # 28  [batch, 224, size/16, size/16]  [40]  # to line 40 # changed from -4 to -5

    [-1, 1, Conv, [23, 1, 1]], # 29
    [-1, 1, DWConv, [23, 3, 1]], # 30
    [-1, 1, Conv, [25, 1, 1]], # 31
    [-1, 1, Conv, [18, 1, 1]], # 32   [batch, 18, size/8, size/8] -> [40] ###
    
    [-5, 1, nn.Upsample, [None, 2, "nearest"]],  # 33   [80]
    [[-1, 12], 1, Concat, [1]],  # 34   [80]  ch = 272      # to line 27  # changed from 11 to 12
    [-1, 1, Conv, [11, 1, 1]], # 35
    [-1, 1, DWConv, [11, 3, 1]], # 36 
    [-1, 1, Conv, [14, 1, 1]], # 37   
    [-1, 1, Conv, [18, 1, 1]], # 38 [batch, 18, 160, 160] -> [80] ###

    [-5, 1, nn.Upsample, [None, 2, "nearest"]],  # 39 [1, 272, 320, 320] -> [160]
    [[-1, 7], 1, Concat, [1]],  # 40  # to line 21
    [-1, 1, Conv, [10, 1, 1]], # 41   
    [-1, 1, DWConv, [10, 3, 1]], # 42 
    [-1, 1, Conv, [13, 1, 1]], # 43   
    [-1, 1, Conv, [18, 1, 1]], # 44   [batch, 18, 320, 320] -> [160]  ###

    [[44, 38, 32], 1, Detect, [nc, anchors]], 


  ]
```
I have no idea about what's going on here, so I just train the model from scratch to get rid of this problem.😞

**Comment by pderrenger** [comment]
@tobymuller233 thank you for providing the detailed model architecture. Since you're encountering NaN values specifically after the nn.Upsample module during forward propagation, this could be related to a few potential issues:

1. Check the input tensor values before the Upsample layers (layers 33 and 39) to verify they are within reasonable ranges. You can add print statements to monitor the tensor statistics:

```python
print(f"Layer input stats: min={x.min().item()}, max={x.max().item()}, mean={x.mean().item()}")
```

2. Consider replacing "nearest" upsampling with "bilinear" mode, which can sometimes be more stable:
```yaml
[-5, 1, nn.Upsample, [None, 2, "bilinear"]]
```

3. Verify the channel dimensions match correctly at the Concat operations following the Upsample layers (layers 34 and 40).

If you're still experiencing issues, you can try adding gradient clipping during training:

```python
python train.py --clip-grad 10.0
```

For more information on debugging NaN issues, please see our guide on [YOLOv5 training tips](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).

**Comment by tobymuller233** [comment]
> @tobymuller233 thank you for providing the detailed model architecture. Since you're encountering NaN values specifically after the nn.Upsample module during forward propagation, this could be related to a few potential issues:
> 
> 1. Check the input tensor values before the Upsample layers (layers 33 and 39) to verify they are within reasonable ranges. You can add print statements to monitor the tensor statistics:
> 
> ```python
> print(f"Layer input stats: min={x.min().item()}, max={x.max().item()}, mean={x.mean().item()}")
> ```
> 
> 2. Consider replacing "nearest" upsampling with "bilinear" mode, which can sometimes be more stable:
> 
> ```yaml
> [-5, 1, nn.Upsample, [None, 2, "bilinear"]]
> ```
> 
> 3. Verify the channel dimensions match correctly at the Concat operations following the Upsample layers (layers 34 and 40).
> 
> If you're still experiencing issues, you can try adding gradient clipping during training:
> 
> ```python
> python train.py --clip-grad 10.0
> ```
> 
> For more information on debugging NaN issues, please see our guide on [YOLOv5 training tips](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).

Thanks for your reply!
I added the log message print code in `yolo.py:_forward_once()`:
![Screenshot 2024-11-18 at 14 29 22](https://github.com/user-attachments/assets/11c7f05c-f5d5-4ec2-b295-0f7741858cc6)
What surprised me is that, the nan value now appears in a random module in the forwarding procedure. I think it's because I'm using hyperparameter evolution. However, the appearance of NaN still reflects some problems.
![Screenshot 2024-11-18 at 14 30 54](https://github.com/user-attachments/assets/5bbeb577-b37f-41e9-871a-2eeb7df0641f)
After checking some webpages or blogs about similar issues, I do BN layer parameter initialization before training:
```python
for m in model.modules():
        if isinstance(m, nn.BatchNorm2d):
            m.reset_running_stats()
            m.reset_parameters()
```
In this piece of code, BN layer parameters are initialized so that the pruned model would not get trouble with BN layers, then what makes the NaN value appear? By the way, you mentioned gradient clipping in your previous answer, but I haven't found the `--clip-grad` option in `train.py`. Also, i don't think this gonna be related to gradient explosion since this is the forwarding procedure.

**Comment by pderrenger** [comment]
It looks like you're experiencing NaN issues potentially due to the hyperparameter evolution process. While your approach to resetting the BatchNorm layers is a good step, NaNs can still arise from other factors such as inconsistent data inputs or numerical instabilities within the model. 

To further troubleshoot, ensure your data preprocessing is robust and check for any outliers or anomalies in your dataset. Since gradient clipping isn't directly applicable here, focus on stabilizing forward computations. You might also want to consider reducing the range of hyperparameters being evolved to see if this mitigates the issue.

If the problem persists, testing the model with a simpler or smaller dataset might provide insights into whether the NaNs are dataset-specific. For additional guidance, reviewing relevant sections in the [YOLOv5 Common Issues Guide](https://docs.ultralytics.com/guides/yolo-common-issues/) might also help.
